documentation/writing_tests.md

```diff
@@ -0,0 +1,949 @@
+    assert.ok(result.value);
+    assert.ok(result.createdAt instanceof Date);
+    assert.ok(result.updatedAt instanceof Date);
+    // And other properties...
+  });
+});
+```
+
+### Test titles
+
+Test files contain `describe` blocks and `it` blocks. The first property that these functions receive will be the title of the group of tests, and the title of each test, respectively for `describe` and `it`. Let's explore these separately:
+
+- The titles sent to the `describe` calls should represent what the collection of its inner test cases (`it` calls) are testing.
+- The titles sent to each `it` call should represent what each specific test case is testing.
+
+Therefore, `describe` titles should be written to reflect something that each of its `it` test cases (or inner `describe` blocks) share in common.
+
+In addition to being expressive, test titles might have specific hashtags (as in, words preceded by the pound sign), to facilitate pattern matching.
+
+#### Pattern matching test titles
+
+`mocha` allows us to only run certain tests as long as they match a given query, with the following arguments:
+
+```
+Test Filters
+  --fgrep, -f   Only run tests containing this string                   [string]
+  --grep, -g    Only run tests matching this string or regexp           [string]
+  --invert, -i  Inverts --grep and --fgrep matches                     [boolean]
+```
+
+To be able to filter tests through pattern matching with any of these filters, the `describe` and `it` titles can have any hashtag, but preferably one or more of the following:
+
+- `#browser` for tests that are only expected to run in the browser.
+- `#node` for tests that are only expected to run on NodeJS.
+- `#live` for tests that should only run live and not be recorded, nor played back.
+
+### Inner parts of a test
+
+While writing test cases, it will be easier to think of the test case's contents as if there were three clear divisions:
+
+- The first lines of the test should deal with setting up the necessary variables and values that will be used for the main functionality that will be tested.
+- The body of the test should focus on using these values to trigger the main functionality that is being tested.
+- The last set of lines should be focused on comparing the results obtained, to confirm that they are in the expected shape, or contain the expected values.
+
+### Using conditionals
+
+Ideally, each test case should only have one possible behavior. To minimize unexpected outcomes, conditionals should be avoided. The only case in which conditionals are encouraged is to specify when to skip a test. Some tests might only run in a browser, or only run in NodeJS, or only run live (no record or playback), for this purpose, conditionals can be used. For example:
+
+```ts
+import { isNode } from "@azure/core-http";
+
+describe("testing the client's basic methods", function() {
+  let client: Client;
+
+  beforeEach(function() {
+    client = new Client();
```

Comment:
No cleanup of client needed? If it's stateless do we need a new one each time?

Reply:
I will move it to "before".

Reply:
So, it depends of where the recorder is defined. Right now, the recorder must be defined in `beforeEach`, since the authentications can mess up playbacks. I've added more context on these beforeEach, as in: a snippet of the definition of the recorder and an afterEach stopping the recorder.

Reply:
We also don't have many clients that require cleanup. Do we? I'm trying to remember.